File: java/core/src/mail/client/ArrivalsMonitorDefault.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */
package mail.client;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import mail.client.model.Direction;

import core.callback.Callback;
import core.callback.CallbackChain;
import core.callback.CallbackDefault;
import core.callback.CallbackWithVariables;
import core.connector.FileInfo;
import core.connector.async.AsyncStoreConnector;
import core.connector.async.AsyncStoreConnectorBase64;
import core.connector.async.AsyncStoreConnectorEncrypted;
import core.connector.async.Lock;
import core.constants.ConstantsStorage;
import core.crypt.Cryptor;
import core.util.LogNull;

public class ArrivalsMonitorDefault extends ArrivalsMonitor
{
	static LogNull log = new LogNull(ArrivalsMonitorDefault.class);
	static final int NUM_FILES_BEFORE_CACHE = 10;
	static final int NUM_FILES_IN_CHECK = 15;
	AsyncStoreConnector store;

	boolean checking = false;
	Object in,out;
	List<FileInfo> files;
	int totalFilesFound=0;
	int numFilesWillCheck=0;
	Lock checkMailLock;
	
	public ArrivalsMonitorDefault (Cryptor cryptor, AsyncStoreConnector connector)
	{
		this.store = new AsyncStoreConnectorEncrypted(cryptor, connector);
		checkMailLock = 
			new Lock(
				new AsyncStoreConnectorBase64(connector), 
				ConstantsStorage.NEW_IN_JSON + "/checkMail.lock", 
				ConstantsStorage.MAIL_CHECK_LOCK_TIME_SECONDS,
				ConstantsStorage.MAIL_CHECK_LOCK_TIME_ALLOWED_BEFORE_RELOCK_SECONDS
			);
	}
	
	public boolean isChecking ()
	{
		return checking;
	}

	@Override
	public void check ()
	{
		if (checking)
			return;
		
		log.debug("check");

		in = null;
		out = null;
		totalFilesFound = numFilesWillCheck = 0;
		
		CallbackChain callbackChain = new CallbackChain()
			.addCallback(master.getEventPropagator().signal_(Events.CheckBegin, (Object[])null))
			.addCallback(master.getCacheManager().update_())
			.addCallback(checkMailLock.lock_())
			.addCallback(check_directory_(ConstantsStorage.NEW_IN_JSON, Direction.IN))
			.addCallback(check_directory_(ConstantsStorage.NEW_OUT_JSON, Direction.OUT))
			.addCallback(check_combine_())
			.addCallback(new CyclicalFileCheck(this))
			.addCallback(check_end_());
		
		// we do not unlock the checkMailLock, because I think this throws
		// too many variables into the equation
		
		checking = true;
		callbackChain.invoke();
	}
	
	public Callback check_end_()
	{
		return 
		new CallbackDefault() {

			@Override
			public void onSuccess(Object... arguments) throws Exception {
				checking = false;
				
				master.getEventPropagator().signal(Events.CheckSuccess, (Object[])null);
				master.getEventPropagator().signal(Events.CheckEnd, (Object[])null);
			}
			
			public void onFailure(Exception e) {
				checking = false;
				
				master.getEventPropagator().signal(Events.CheckFailure, "Failed");
				master.getEventPropagator().signal(Events.CheckEnd, (Object[])null);
			}
		};	
	}
	
	public Callback check_directory_ (String path, Direction direction)
	{
		log.debug("check directory", path);
		
		return store.list_(path + "/").addCallback(check_accumulate_(direction));
	}
	
	public Callback check_accumulate_ (Direction direction)
	{
		return new CallbackDefault(direction) {
			public void onSuccess(Object... arguments) throws Exception {
				Direction direction = V(0);
				if (direction == Direction.IN)
					in = arguments[0];
				else
					out = arguments[0];
				
				next();
			}
		};
	}
	
	public Callback check_combine_ ()
	{
		return new CallbackDefault() {
			public void onSuccess(Object...arguments) throws Exception {

				log.debug("check_combine");
				checkMailLock.testLock((List<FileInfo>)in);

				files = new ArrayList<FileInfo>();
				if (in instanceof List)
				{
					List<FileInfo> lin = (List<FileInfo>)in;
					log.debug("check_combine","in", lin.size());
					
					for (FileInfo file : lin)
					{
						if (file.path.endsWith(".lock"))
							continue;
						
						if (master.getArrivalsProcessor().alreadyProcessed(file.path))
							continue;
						
						log.debug("found file",file);
						
						files.add(file);
						file.user = Direction.IN;
					}
				}
				if (out instanceof List)
				{
					List<FileInfo> lout = (List<FileInfo>)out;
					log.debug("check_combine","out", lout.size());
					for (FileInfo file : lout)
					{
						if (master.getArrivalsProcessor().alreadyProcessed(file.path))
							continue;
						
						files.add(file);
						file.user = Direction.OUT;
					}
				}
			
				totalFilesFound = files.size();
				
				Collections.sort(files, new FileInfo.SortByDateAscending());
				List<FileInfo> segment = files.subList(0, Math.min(files.size(), NUM_FILES_IN_CHECK));
				files = segment;
				
				numFilesWillCheck = files.size();
				
				log.debug("check_combine","final", files.size());
				next();
			}
		};
	}
}

File: cpp/core/mailiverse/mail/manager/ArrivalsMonitorDefault.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "ArrivalsMonitorDefault.h"
#include "mailiverse/utilities/Thread.h"
#include "mailiverse/utilities/Log.h"
#include "mailiverse/utilities/Strings.h"
#include "Master.h"
#include "Events.h"
#include "Constants.h"

using namespace mailiverse::mail::manager;
using namespace mailiverse::core::connector;
using namespace mailiverse::core::crypt;
using namespace mailiverse;
using namespace mailiverse::utilities;

ArrivalsMonitorDefault::ArrivalsMonitorDefault(Cryptor *cryptor, StoreConnector *connector) :
	store(new core::connector::StoreConnectorEncrypted(cryptor, connector)),
	finished(false),
	checking(false),
	checkMailLock(
		connector, Constants::MAIL_IN_JSON_PREFIX + Constants::CHECK_MAIL_LOCK_FILE_NAME,
		Constants::CHECK_MAIL_LOCK_TIME_SECONDS, 
		Constants::CHECK_MAIL_LOCK_TIME_ALLOWED_BEFORE_RELOCK_SECONDS
	)
{
}

ArrivalsMonitorDefault::~ArrivalsMonitorDefault()
{
}

void ArrivalsMonitorDefault::markFinished ()
{
	finished = true;
}

void ArrivalsMonitorDefault::check ()
{
	try
	{
		checking = true;
		getMaster()->getCacheManager()->update();
		doCheck();
	}
	catch (Exception &e)
	{
		getMaster()->getEventPropagator()->signal(Events::CheckFailure, NULL);
		checking = false;
		throw e;
	}
}

bool ArrivalsMonitorDefault::isChecking ()
{
	return checking;
}

Vector<ArrivalsMonitorDefault::Listing> ArrivalsMonitorDefault::getListing ()
{
	StoreConnector::FileListing in = store->list(Constants::MAIL_IN_JSON_PREFIX, getMaster()->getArrivalsProcessor()->getRequestDateRange());
	StoreConnector::FileListing out = store->list(Constants::MAIL_OUT_JSON_PREFIX, getMaster()->getArrivalsProcessor()->getRequestDateRange());

	checkMailLock.testLock(in);

	Vector<Listing> result;
	for (auto &i : in)
	{
		if (endsWith(i.path, ".lock"))
			continue;
	
		Listing l;
		l.file = i;
		l.direction = Direction::IN;
		l.file.path = l.file.path;
		result.add(l);
	}

	for (auto &i : out)
	{
		Listing l;
		l.file = i;
		l.direction = Direction::OUT;
		l.file.path = l.file.path;
		result.add(l);
	}

	std::sort(result.begin(), result.end(), ListingComparator());
	return result;
}

void ArrivalsMonitorDefault::doCheck ()
{
	ArrivalsProcessor *arrivalsProcessor = getMaster()->getArrivalsProcessor();

	checking = true;
	getMaster()->getEventPropagator()->signal(Events::CheckBegin, NULL);
	
	checkMailLock.relock();
	auto listing = getListing();

	int totalOutstanding = 0;
	List<Listing> itemsToCheck;
	for (auto &i : listing)
	{
		if (!arrivalsProcessor->alreadyProcessed(i.file.path))
		{
			totalOutstanding++;
			
			if (totalOutstanding <= Constants::NUM_ITEMS_TO_CHECK_IN_ONE_CYCLE)
				itemsToCheck.push_back(i);
		}
	}
	
	int numChecked = 0;
	for (auto &i : itemsToCheck)
	{
		if (finished)
			break;

		checkMailLock.relock();
		
		numChecked++;

		arrivalsProcessor->getMaster()->getEventPropagator()->signal(
			Events::CheckStep, 
			utilities::newArg<StringPtr>(
				new String(toString(numChecked) + ":" + toString(itemsToCheck.size()) + ":" + toString(totalOutstanding))
			)
		);

		try
		{
			LogDebug (mailiverse::mail::manager::ArrivalsMonitor, "processing " << i.file.path);

			arrivalsProcessor->processSuccess (
				i.direction, i.file.path, i.file.date, store->get(i.file.path).first
			);
		}
		catch (Exception &e)
		{
			LogDebug (mailiverse::mail::manager::ArrivalsMonitor, "processing failing! " << i.file.path << " what " << e.what());

			arrivalsProcessor->processFailure (
				i.direction, i.file.path, i.file.date, e
			);
		}

/*
		if (++numChecked % NUM_PROCESSED_FLUSH == 0)
			getMaster()->getCacheManager()->flush();
*/
	}
	
	getMaster()->getEventPropagator()->signal(Events::CheckEnd, NULL);
	checking = false;
}
